src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,293 @@
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.ConvertToSwitch;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertToSwitch
+{
+    internal sealed class CSharpConvertToSwitchCodeRefactoringProvider : AbstractConvertToSwitchCodeRefactoringProvider
+    {
+        private abstract class Pattern
+        {
+            public sealed class ByValue : Pattern
+            {
+                public readonly ExpressionSyntax Expression;
+                public ByValue(ExpressionSyntax expression)
+                {
+                    Expression = expression;
+                }
+            }
+            public sealed class ByType : Pattern
+            {
+                public readonly TypeSyntax Type;
+                public readonly SyntaxToken Identifier;
+                public ByType(TypeSyntax type, SyntaxToken identifier)
+                {
+                    Type = type;
+                    Identifier = identifier;
+                }
+            }
+            public sealed class Discarded : Pattern
+            {
+                public readonly TypeSyntax Type;
+                public Discarded(TypeSyntax type)
+                {
+                    Type = type;
+                }
+            }
+            public sealed class Guarded : Pattern
+            {
+                public readonly Pattern Pattern;
+                public readonly ExpressionSyntax Guard;
+                public Guarded(Pattern pattern, ExpressionSyntax guard)
+                {
+                    Pattern = pattern;
+                    Guard = guard;
+                }
+            }
+        }
+
+        private static IEnumerable<ExpressionSyntax> GetLogicalOrExpressionOperands(ExpressionSyntax syntaxNode)
+        {
+            syntaxNode = syntaxNode.WalkDownParentheses();
+            while (syntaxNode.IsKind(SyntaxKind.LogicalOrExpression))
+            {
+                var binaryExpression = (BinaryExpressionSyntax)syntaxNode;
+                yield return binaryExpression.Right.WalkDownParentheses();
+                syntaxNode = binaryExpression.Left.WalkDownParentheses();
+            }
+
+            yield return syntaxNode;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var ifStatement = root.FindNode(context.Span) as IfStatementSyntax;
```

Comment:
i think, ideally, you would see this if you cursor was in the 'header' of the 'if'.  i.e. from the 'if' to the end of the condition.